core/vm/eof.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2021 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+)
+
+type eofVersion int
+type headerSection byte
+
+const (
+	eofFormatByte  byte          = 0xEF
+	eof1Version    eofVersion    = 1
+	kindTerminator headerSection = 0
+	kindCode       headerSection = 1
+	kindData       headerSection = 2
+	kindType       headerSection = 3
+
+	versionOffset      int = 2
+	sectionHeaderStart int = 3
+)
+
+var eofMagic []byte = []byte{0xEF, 0x00}
+
+// hasEOFByte returns true if code starts with 0xEF byte
+func hasEOFByte(code []byte) bool {
+	return len(code) != 0 && code[0] == eofFormatByte
+}
+
+// hasEOFMagic returns true if code starts with magic defined by EIP-3540
+func hasEOFMagic(code []byte) bool {
+	return len(eofMagic) <= len(code) && bytes.Equal(eofMagic, code[0:len(eofMagic)])
+}
+
+// isEOFVersion1 returns true if the code's version byte equals eof1Version. It
+// does not verify the EOF magic is valid.
+func isEOFVersion1(code []byte) bool {
+	return versionOffset < len(code) && code[versionOffset] == byte(eof1Version)
+}
+
+type EOF1Header struct {
+	typeSize uint16   // Size of type section. Must be 2 * n bytes, where n is number of code sections.
+	codeSize []uint16 // Size of code sections. Cannot be 0 for EOF1 code. Equals 0 for legacy code.
+	dataSize uint16   // Size of data section. Equals 0 if data section is absent in EOF1 code. Equals 0 for legacy code.
+}
+
+func NewEOF1Header(code []byte, jt *JumpTable, validated bool) (EOF1Header, error) {
```

Comment:
I think this (and the other ) constructors are not great. 
- The `code` is not the actual code, it's the 'blob' of data which comprises both header + code sections. 
- This maybe should be named `UnmarshalEOF1Header`, because what it does is try to create an eof1 header based off the marshalled content in `code`. 

So I think it would be suitable to rename this (and also add doc-string while at it).
Also, to add new constructor(s), suitable for creation for other purposes. 


For example, the tests want to create "an eof header with a single code segment with the code `0x0a0b`". RIght now, they do basically
```
	code := common.FromHex("EF000101000100" + "0C")
	header, err := NewEOF1Header(code, jt, false)
```
This is clunky - a better api would allow to do this: 
```
header, err := NewEofHeader(nil, nil, []byte{0c})
```
I don't know exactly what arguments would be needed, and which would have to be variadic, but you probably see what I'm getting at. 

And the same goes for the container type.

Reply:
Example function that could help 
```golang
// EofBlobFromCode creates an EOFv1 header blob.
func EofBlobFromCode(byteCode []byte, data []byte) []byte {
	headerSize := 3 + 3 + 1 // magic, codesec, terminator
	if data != nil {
		headerSize += 3
	}
	blob := make([]byte, headerSize+len(byteCode)+len(data))
	blob[0] = 0xEF
	blob[1] = 0x00
	blob[2] = 0x01
	// Add a code section
	blob[3] = 1 // code
	blob[4] = byte(uint16(len(byteCode)) >> 8)
	blob[5] = byte(uint16(len(byteCode)))
	if size := len(data); size > 0 {
		// Add a data section
		blob[6] = 2 // data
		blob[7] = byte(uint16(size) >> 8)
		blob[8] = byte(uint16(size))
		copy(blob[headerSize+len(byteCode):], data)
	}
	blob[headerSize] = 0x00 // terminator
	copy(blob[headerSize:], byteCode)
	return blob
}
```
Then the tests could use that method, 
```golang
func TestEofBlob(t *testing.T) {
	have := EofBlobFromCode(common.FromHex("0x0c00"), nil)
	want := common.FromHex("EF0001010002000C00")
	if !bytes.Equal(have, want) {
		t.Fatalf("error:\nhave %x\nwant %x", have, want)
	}
}
```
And we could avoid the `EF0001010002000C00` strings which are a bit opaque for someone not fully versed in it